### DISCUSSION TITLE
EmonCMS Integration

### DISCUSSION THREAD

**Discussion by blaal02** [Open-Building-Management/emoncms#23] [Emoncms integration in home assistant]
This EmonCMS implementation is great. Working nicely for me, logging from EmonTXs, EmonTHs (via EmonPi) & an IotaWatt.
But I'm curious, what others are doing with the Home Assistant integration? The core version is "working", but has that issue with Unique Ids.

**Comment by alexandrecuer** [comment]
Thanks nice to see it is usefull

You mean the emoncms Home assistant integration ? Cf 
https://www.home-assistant.io/integrations/emoncms/

I know the integration needs updating :
https://github.com/home-assistant/core/pull/86840

**Comment by blasa01** [comment]
Exactly!
There seemed to be some politics involved in getting the updates into the core integration (There are some PRs that have been created, but much discussion aound how to create a "Unique Id")
So, I was wondering if anyone here had found (or created) a custom fork that had some of that cleanup working.
Thanks, Sandy

**Comment by alexandrecuer** [comment]
Just tested the `emoncms` integration which is not very much used (9 active installations)

![image](https://github.com/Open-Building-Management/emoncms/assets/24553739/22e2feeb-b458-4968-b8cc-c840da328cdc)

From my side, it is working ...

Just catch the addon IP through emoncms admin module, in my case it is 172.30.33.0
![image](https://github.com/Open-Building-Management/emoncms/assets/24553739/d0dd8151-d4ca-4533-962d-11344507090d)
Then added the following in configuration.yaml and restarted HA :
```
sensor:
  - platform: emoncms
    api_key: XXXXXXXXXXXXXXX
    url: http://172.30.33.0
    id: 1
```
after you see the feeds in settings > states, here `sensor.emoncms_outdoor_humidity` and `sensor.emoncms_outdoor_temp` 
![image](https://github.com/Open-Building-Management/emoncms/assets/24553739/dfb7fd05-9a03-41b2-89e4-94e1290e4feb)

I have not yet tested the `emoncms history` integration which seems to be a little more used....

**Reply by blasa01** [reply]
That's the one I'm using, and indeed it does work. But (as you pointed out), it's not used much, and has those Unique Id issues (so you can't do anything with the feed settings).
So, I was just wondering what others (like you!) are using to integrate EmonCMS data in Home Assistant?
![image](https://github.com/Open-Building-Management/emoncms/assets/161655561/c2cf9047-9092-4823-8128-557c46404167)

**Reply by alexandrecuer** [reply]
Well, I am really a beginner in Home Assistant usage. I just did the addon because I was working on a docker image for emoncms and it was really easy to recycle it as an addon. 

When emoncms dashboard are not enough, I use a small python app with [d3.js](https://d3js.org/)
![image](https://github.com/Open-Building-Management/emoncms/assets/24553739/e8f67dd2-0b19-4578-87f8-db9c4f088c6d)

You Energy graph is very nice,  all the datas come from emoncms ?
 
Could you explain more what is this unique Id problem ? 

To what i've understood, an integration must have 2 sides, one external library on pypi which interacts with the hardware (through local polling here) and a wraper in home assistant to convert the data to the appropriate format. This is not the case in the emoncms integration and so that is why the PR are not merged....

**Reply by blasa01** [reply]
Energy graph - yes, I just added all my kWh sensors from the default EmonCMS integration as "Individual devices". All I did was changed the name to get rid of the ":EmonCMS..." decoration.
![image](https://github.com/Open-Building-Management/emoncms/assets/161655561/bf6fd6ef-d3a2-4894-ae9c-b28adf2c6388)
Unique Id - If you try to edit any of the EmonCMS sensors from the UI, you get:
![image](https://github.com/Open-Building-Management/emoncms/assets/161655561/f5425557-ca2d-4926-9f5c-b781d0cc1e03)
Where "politics" comes in is that the HA developers insist that the assigned id be somehow hardware assigned & not user controllable - but feeds are user created & configured, so concept doesn't really exist.
And yes, I too usually use EmonCMS graphs for detailed views:
![image](https://github.com/Open-Building-Management/emoncms/assets/161655561/188c44c1-d1b9-4547-abef-5837ec410af6)
Home Assistant "Beginner" - not from what I'm seeing!!

**Reply by alexandrecuer** [reply]
Thanks for the clarification 

emoncms feed number is unique, so there should not be any problem I think

The integration already involve a `get_id` method producing a unique id based on feed tag, name and id. 

The only thing to do in order for feeds edition to work is to rename `self._name` to `self._attr_name` and `self._identifier` to `self._attr_unique_id` so that then can be interpreted by HA as friendly name and unique id

So I've submitted a small PR https://github.com/home-assistant/core/pull/117911 

If accepted, it should solve the pb

**Reply by alexandrecuer** [reply]
@blasa01 : usually, you need more than one emoncms instance integrated in ha or one enough as sync module permis to share feeds between emoncms modules ? 

@borpin: you know the common practice if there is one among emoncms users ?

**Reply by blasa01** [reply]
In my case (and I'm guessing most) a single EmonCMS instance is used to integrate inputs from multiple sources (EmonTX's, EmonTH's & IotaWatt's). I can't come up with a use case for multiple EmonCMS's feeding into a single HA - at least not in a single residential setting. When we get to commercial or multi-home implementations, there might be some purpose?

**Reply by alexandrecuer** [reply]
@blasa01 : Ok understood

I have a few questions  : 

1) why do yu need to be able to edit feed on home-assistant ? If things are correctly tagged in emoncms, the same classification should be relevant in home-assistant ? Maybe you need the ability to disable some feeds in order they dont appear on the dashboard ?

2) in case of migrating the integration to the configflow process the HA team is promoting : 
-  `sensor_names` and `unit_of_measurement` could be probably removed as feed editions through the UI will provide something similar...
- are `include_only_feed_id`, ` exclude_feed_id` and  `value_template` really usefull ?

**Comment by blasa01** [comment]
Hi Alexandre,

  1.  Correct, unnecessary ones I’d like to disable to reduce overhead. The rest I’d like to hide (not visible) so they don’t appear in the default dashboards.
  2.  They’re trying to volunteer you to do everything!
     *   Sensor names – I don’t know enough about ‘feed editions’ to comment. I did rename my feeds in the Energy tab config. They imported as (e.g.) ‘EmonCMS CT 4 Office’ in Energy I renamed to just ‘Office’. Not sure where the EmonCMS piece came from, it’s not capitalized that way anywhere in my config. CT4 is part of the feed name I created – so I could have changed the name in EmonCMS.
     *   Unit of Measure – theoretically these should match & not need to be set. However, I do have one issue where EmonCMS is configured a ‘Gal’, and I think HA is looking for  ‘gal’ (trying to add water consumption to the energy tab)
     *   Include/exclude – if I can disable feeds later it’s not too important. But if it could support masks it would be nice – e.g. include “*kWh” (I don’t see the instantaneous W readings being useful in HA)

Thanks,

Alexander (Sandy) Blackburn

**Comment by borpin** [comment]
The clarity is required about the Integration (a means of HA accessing emonCMS data) and the add-on (an instance of emonCMS running inside HA environment.

The Integration does need work, thanks @alexandrecuer for that as it is outside my skill set.

My personal preference would be to move the integration out of 'Core' so we don't get caught up in the Devs politics and move it to an HACS custom install - Trystan has setup an empty repo for that.

It then needs a UI setup to be created which could pull in feed names such that they can just be selected.

**Comment by alexandrecuer** [comment]
work in progress here for adding a uuid to emoncms : https://github.com/emoncms/emoncms/pull/1888

**Comment by alexandrecuer** [comment]
Done in https://github.com/home-assistant/core/releases/tag/2024.10.0b0
